### PR TITLE
Add basic section editing buttons to MDN

### PR DESCRIPTION
**Pull request by darkwing**
Not ready but a rough prototype for "section editing" whereby:

1.  Finds headings, adds "edit" links
2.  Loads edit page, scrolls down to the relevant section.

### CODE REVIEW

**Comment by darkwing** [comment]
@stephaniehobson for CSS review
@groovecoder for functionality review

**Comment by groovecoder** [comment]
Works well. Couple additions:

* [ ] Add Google Analytics event tracking to clicks on the `section-edit` links, so we can measure popularity and usage
* [ ] Add Google Analytics custom dimension when the waffle flag is active (`ga('set', 'dimension9', 'Enabled');`), so we can segment and analyze visitors who saw the feature (but may not have used it)

**Comment by groovecoder** [comment]
Also, needs a bug number in the commit message. Use bug 839559 since this implements [comment 56](https://bugzilla.mozilla.org/show_bug.cgi?id=839559#c56).

**Comment by stephaniehobson** [comment]
Can we restrict it to `$('#wikiArticle').children('h2')`?

A few use cases that would solve:

screen shot 2015-09-21 at 19 45 30
---
screen shot 2015-09-21 at 19 44 27
---
screen shot 2015-09-21 at 20 39 55
---
screen shot 2015-09-21 at 19 43 33

It removes it from a few places where it would be useful but is overall an improvement I think.

**Comment by stephaniehobson** [comment]
Actually, this is a branch, isn't it? I'll do it, and add some CSS tweaks.

**Comment by darkwing** [comment]
@stephaniehobson ftw

**Comment by groovecoder** [comment]
This got its rebase but it's about to have its 2-week :birthday:. Can we merge it for its :gift: ?

**Comment by stephaniehobson** [comment]
I can't review this one, I wrote code for it.

**Comment by groovecoder** [comment]
I got this trying to run `./scripts/compile-stylesheets`:
```
Error: ./scripts/../kuma/static/stylus/wiki.styl:67:10
   63| @require 'components/wiki/content/moreinfo';
   64| @require 'components/wiki/content/summary';
   65| @require 'components/wiki/customcss';
   66| @require 'components/wiki/open-in-host';
   67| @require 'components/wiki/section-edit';
----------------^
   68|
   69|
   70| /*

failed to locate @require file components/wiki/section-edit.styl
```

**Comment by darkwing** [comment]
Could be a rebase issue, I'll check it out

**Comment by darkwing** [comment]
Updated, sorry @groovecoder -- put the file in the wrong place -- wont happen again.

File: kuma/static/stylus/components/wiki/section-edit.styl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    margin-left: 20px;
+    set-smaller-font-size();
+    opacity: 0;
+
+    @media $media-query-mobile {
+        display: none;
+    }
+
+    &:focus,
+    h2:hover & {
```
Comment (groovecoder) [inline kuma/static/stylus/components/wiki/section-edit.styl]: Let's start with this. If it's effective, we might like to do a follow-on experiment to make the edit icons always visible on the section headers. Maybe just for signed-in users?

The goal is to make readers more aware that they can edit MDN, and to make it easier for them to do so.

**Comment by groovecoder** [comment]
I think I'm getting the old bug we saw in Madison? i.e.,
1. Click section edit link
2. Click in the content in the section
   * The CKEditor box scrolls back up to the top

**Comment by groovecoder** [comment]
section-scroll-back

**Comment by darkwing** [comment]
OK, so the only way to make the "jump" not happen was to trigger autogrow upon page load and not upon page focus.  We'll see how the writers like that.

Additionally, to get the cursor to the right position in Chrome, I've added a range snippet.

**Comment by groovecoder** [comment]
Much better now. There are certainly some follow-ups we can make, but this is a good first experiment.

File: kuma/wiki/templates/wiki/document.html [hunk 2]
```diff
@@ -429,4 +429,35 @@ <h2 class="offscreen">{{ _('Document Tags and Contributors') }}</h2>
...
+                .append($('<span>'))
+                .find('span')
+                .text(gettext('Edit'))
+                .end()
+                .on('click', function(e) {
+                    e.preventDefault();
+
+                    mdn.analytics.trackEvent({
+                        category: 'Section Edit',
+                        action: href
```
Comment (groovecoder) [inline kuma/wiki/templates/wiki/document.html]: Could we change this to `action: $(this).attr('id')`? That way in GA we'll see actions like "#Browser compatibility", "#Specifications", "#Examples" so we can learn which sections of pages are most popular for edits/contributions.

Comment (darkwing) [inline kuma/wiki/templates/wiki/document.html]: Yeah, sure

**Comment by darkwing** [comment]
Updated!

**Comment by groovecoder** [comment]
LGTM.

**Comment by a2sheppy** [comment]
Limiting to <h2> is problematic because <h2> sections can be very, very long at times and really limits the usefulness of this feature. I don't see the problem with the screenshot shobson posted. Is it a technical problem or some kind of UX concern? I don't find the edit buttons problematic, especially since they appear and disappear as you mouse on and off of each heading.

While I'm at it, I've noticed that at least sometimes, the heading is scrolled to just off the bottom edge of the edit box, instead of to the top of it.

**Comment by stephaniehobson** [comment]
I see it has a UX concern to have them showing up where they wreck formatting (community box) or within about 100px of each other because it clutters the interface and looks silly. I did make that call when they were visible though (as they are in the screen shot).

I'm willing to revisit if we stop talking about making them visible 100% of the time but that conversation is still on going as well.

**Comment by a2sheppy** [comment]
I personally prefer having them not visible all the time. Makes the page more readable in general.

Another option for section editing -- what if instead of buttons, we added an "edit this section" link to the context menu on the heading? That would avoid UX clutter entirely.

**Comment by groovecoder** [comment]
+1 making them invisible until hover. And/or we could run some follow-up experiments if/when we want to encourage more edits/sign-ups. E.g., only show them to users who are signed-in?

**Comment by groovecoder** [comment]
I guess +1 **keeping** them invisible until hover.

**Pull request merged**